File: leetcode1038/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    /// DFS

    /// Time: O(n) | Space: O(1)
    /// where n is the number of nodes in the tree
public:
    TreeNode* bstToGst(TreeNode* root) {
        int acc = 0;
        traverse(&acc, root);

        return root;
    }

    // In-order traversal and traverse the right node first instead of left
    void traverse(int* acc, TreeNode* root) {
        if (root != NULL) {
            traverse(acc, root->right); // traverse right
            root->val += *acc; // update root
            *acc = root->val; // update acc
            traverse(acc, root->left); // traverse left
        }
    }
};

File: leetcode1038/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    /// DFS
    /// 
    /// Time: O(n) | Space: O(n)
    /// where n is the number of nodes in the tree
    pub fn bst_to_gst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut acc: i32 = 0;

        Self::traverse(&mut acc, root)
    }

    // In-order traversal and traverse right first instead of left
    fn traverse(acc: &mut i32, root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        if let Some(ref node) = root {
            let mut node = node.borrow_mut();
            Self::traverse(acc, node.right.clone()); // traverse right first
            node.val += *acc; // update the node value
            *acc = node.val; // update the accumulator
            Self::traverse(acc, node.left.clone()); // traverse left last
        }

        root
    }
}